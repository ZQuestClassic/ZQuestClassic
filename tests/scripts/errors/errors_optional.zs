// ok
int optional_params_1(int a = 0) {
	return a;
}

// ok
int optional_params_2(int a = 0, int b = 1 + 1) {
	return a;
}

// not ok, strings currently aren't considered compile time values (but probably should be)
int optional_params_3(char32[] a = "") {
	return a;
}

// this should be not ok, b/c it is not a compile time value
// But there is no error right now...
// `DataType const* getType = (*it)->getReadType(scope, this);` in `RegistrationVisitor::caseFuncDecl` is somehow returning NULL
// for this
// compiled function doesn't work at all.
int optional_params_4(int a = Player->X) {
	return a;
}

// not ok, not a compile time value
int optional_params_5(int a = optional_params_1()) {
	return a;
}

void fn() {
	optional_params_1();
	optional_params_1(1);
	optional_params_2();
	optional_params_2(1);
	optional_params_2(1, 2);
	optional_params_3();
	optional_params_4();
	optional_params_5();
}
