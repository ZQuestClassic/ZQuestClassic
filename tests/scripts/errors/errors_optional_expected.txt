stderr:

Compiling 'errors_optional.zs'
Pass 1: Parsing
Pass 2: Preprocessing
Pass 3: Registration
errors_optional.zs Line 12 @ Columns 36-38 - Error S109: Function 'optional_params_3' has an optional parameter whose default value is not a constant expression

10    
11    // not ok, strings currently aren't considered compile time values (but probably should be)
12    int optional_params_3(char32[] a = "") {
                                         ^~


errors_optional.zs Line 26 @ Columns 31-48 - Error R075: Global variable initializers can only contain calls to 'constexpr' functions

24    
25    // not ok, not a compile time value
26    int optional_params_5(int a = optional_params_1()) {
                                    ^~~~~~~~~~~~~~~~~


errors_optional.zs Line 26 @ Columns 31-50 - Error S109: Function 'optional_params_5' has an optional parameter whose default value is not a constant expression

24    
25    // not ok, not a compile time value
26    int optional_params_5(int a = optional_params_1()) {
                                    ^~~~~~~~~~~~~~~~~~~


Failure!
Compile finished with exit code '109' (compiled with errors)

stdout:

{
  "success": false,
  "code": 109,
  "diagnostics": [
    {
      "range": {
        "start": {
          "line": 11,
          "character": 35
        },
        "end": {
          "line": 11,
          "character": 37
        }
      },
      "severity": 1,
      "message": "S109: Function 'optional_params_3' has an optional parameter whose default value is not a constant expression"
    },
    {
      "range": {
        "start": {
          "line": 25,
          "character": 30
        },
        "end": {
          "line": 25,
          "character": 47
        }
      },
      "severity": 1,
      "message": "R075: Global variable initializers can only contain calls to 'constexpr' functions"
    },
    {
      "range": {
        "start": {
          "line": 25,
          "character": 30
        },
        "end": {
          "line": 25,
          "character": 49
        }
      },
      "severity": 1,
      "message": "S109: Function 'optional_params_5' has an optional parameter whose default value is not a constant expression"
    }
  ],
  "metadata": {
    "currentFileSymbols": 6,
    "symbols": 0,
    "identifiers": 0,
    "elided": true
  }
}
